internal/controller/certificaterequests/apply_test.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificaterequests
+
+import (
+	"strconv"
+	"sync"
+	"testing"
+
+	fuzz "github.com/google/gofuzz"
+	"github.com/stretchr/testify/assert"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+)
+
+// This test ensures that when a Certificate object is serialized in
+// preparation for a Certificate Apply call. Only object meta/type and spec
+// field should be present.
+func Test_serializeApply(t *testing.T) {
+	// Expected serialized Certificate Apply object. Should only contain base
+	// type object, object meta object, and spec. status should be empty. Empty
+	// spec should be deterministic.
+	const (
+		expReg      = `^{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{.*},"spec":{.*},"status":{}}$`
+		expEmptyReg = `^{"kind":"CertificateRequest","apiVersion":"cert-manager.io/v1","metadata":{.*},"spec":{"issuerRef":{"name":""},"request":null},"status":{}}$`
+		numJobs     = 10000
+	)
+
+	var wg sync.WaitGroup
+	jobs := make(chan int)
+
+	wg.Add(numJobs)
+	for i := 0; i < 3; i++ {
+		go func() {
+			for j := range jobs {
+				t.Run("fuzz_"+strconv.Itoa(j), func(t *testing.T) {
+					var req cmapi.CertificateRequest
+					fuzz.New().NilChance(0.5).Fuzz(&req)
+
+					// Test regex with non-empty spec.
+					reqData, err := serializeApply(&req)
+					assert.NoError(t, err)
+					assert.Regexp(t, expReg, string(reqData))
```

Comment:
Thought: perhaps here we could compare unmarshaled spec of `expReg` with the fuzzed `req.Spec`? That way we could test that the spec fields were preserved correctly (although it would, to an extent be testing that `json.Marshal`/`json.Unmarshal` works)

Reply:
So effectively testing the roundtrip of the resource through the serializer? I think that makes sense to me- will add it in